### PR TITLE
feat: Add CircleCI integration

### LINKED ISSUE #1957: [CircleCI] Base
### Description

Create a **SuperPlane integration for CircleCI** and **two starter components** (one trigger, one action) so users can build workflows with CircleCI in SuperPlane. This issue covers the base integration and the first components listed below.

**Link**: https://circleci.com/

### Connection Method

Part of this task is exploring and determining the best approach to connect CircleCI with SuperPlane.

**Suggested starting point - API Token**: Connect using a CircleCI API token
* Create token: CircleCI → User Settings → Personal API Tokens
* Token is used for CircleCI API (trigger pipelines, get status, etc.)

Consider the available authentication options and evaluate which method provides the best balance of security, ease of setup, and functionality for our use cases.

For inspiration, check the existing **Semaphore integration** implementation as a reference for how similar CI/CD platform integrations are structured.

### First Components

‼️ **IMPORTANT**

Creating at least 1 `trigger` and 1 `action` component to test out the Base is required for completing this issue.

#### 1. On Pipeline Completed (Trigger)

A trigger that listens to CircleCI webhook events when pipelines complete. This component passively receives completion notifications from CircleCI and emits events to start SuperPlane workflow executions.

#### 2. Trigger Pipeline (Action)

An action that starts a pipeline in CircleCI and tracks its execution until completion. The component should monitor the pipeline status (via webhook and/or polling as fallback) and emit the final result (success/failure) only once the pipeline has finished. See Semaphore's `Run Workflow` action for reference on how to implement execution tracking with webhook + polling.

### Bounty Details

* Read the [Contributing guide](https://github.com/superplanehq/superplane/blob/main/docs/contributing) and [Integration bounties](https://github.com/superplanehq/superplane/blob/main/docs/contributing/bounties.md) before starting.
* We encourage you to join [Discord](https://discord.gg/superplane) and reach out to the team with any questions.
* Your PR must include a **short description** of the changes.
* Attach a **short video** showing: (1) setting up the integration, and (2) the working components in use.

### Acceptance Criteria (for SuperPlane team)

- [ ] integration works
- [ ] starter components work
- [ ] has proper tests
- [ ] has proper documentation
- [ ] passed code quality review
- [ ] passed functionality review
- [ ] passed ui/ux review

---

⚠️ **DISCLAIMER**: Pull requests submitted without a video demonstration will be automatically closed.

### PR DESCRIPTION
**Pull request by vikramships**
## What Changed

Added CircleCI integration with API token authentication. This includes two components: a trigger that listens for pipeline completions and an action that starts pipelines and tracks them to completion.

## Why

This integration enables SuperPlane users to orchestrate CircleCI pipelines within their workflows. Users can trigger builds, monitor pipeline status, and react to completion events without leaving SuperPlane.

## How

**Base Integration:**
- Uses CircleCI Personal API Token for authentication
- Implements webhook provisioning for real-time events
- Follows the Semaphore integration pattern

**On Pipeline Completed (Trigger):**
- Registers webhooks with CircleCI to receive workflow completion events
- Verifies webhook signatures using HMAC SHA256
- Emits events when workflows finish (success/failed/canceled)

**Trigger Pipeline (Action):**
- Starts pipelines via CircleCI API with configurable parameters
- Monitors completion using webhook + polling fallback (10s interval)
- Routes to success/failed output channels based on final workflow status
- Automatically injects SuperPlane execution context as pipeline parameters

**Tests & Documentation:**
- Unit tests for all components
- Integration tests for webhook handling
- Full documentation in `docs/components/CircleCI.mdx`

## Demo Video

https://www.youtube.com/watch?v=PvzHGetsujk

The video shows:
1. Integration setup with API token
2. On Pipeline Completed trigger detecting a pipeline completion
3. Trigger Pipeline action starting a pipeline and tracking it to success

## Related Issues

Closes #1957

## Breaking Changes

None

### CODE REVIEW

File: pkg/integrations/circleci/list_resources.go [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package circleci
+
+import (
+	"github.com/superplanehq/superplane/pkg/core"
+)
+
+func (c *CircleCI) ListResources(resourceType string, ctx core.ListResourcesContext) ([]core.IntegrationResource, error) {
+	// CircleCI doesn't have a concept of listable resources like projects
+	// Users need to manually enter their project slug
+	return []core.IntegrationResource{}, nil
```
Comment (shiroyasha) [inline pkg/integrations/circleci/list_resources.go]: It seems that you can list available projects based on the me endpoint. 

Take a look at this: https://support.circleci.com/hc/en-us/articles/360047534371-Find-all-Organization-Projects-via-the-CircleCI-API

File: pkg/integrations/circleci/on_pipeline_completed.go [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+
+func (p *OnPipelineCompleted) Configuration() []configuration.Field {
+	return []configuration.Field{
+		{
+			Name:        "projectSlug",
+			Label:       "Project Slug",
+			Type:        configuration.FieldTypeString,
+			Required:    true,
+			Description: "CircleCI project slug (e.g., gh/username/repo)",
+			Placeholder: "gh/username/repo",
```
Comment (shiroyasha) [inline pkg/integrations/circleci/on_pipeline_completed.go]: If possible to list projects, lets use a list resource for this. It will render a select box in the application, instead of requiring the user to find and enter the slug.

File: pkg/integrations/circleci/on_pipeline_completed.go [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+	// Parse webhook payload
+	data := map[string]any{}
+	err = json.Unmarshal(ctx.Body, &data)
+	if err != nil {
+		return http.StatusBadRequest, fmt.Errorf("error parsing request body: %v", err)
+	}
+
+	// Verify this is a workflow-completed event
+	eventType, ok := data["type"].(string)
+	if !ok || eventType != "workflow-completed" {
```
Comment (shiroyasha) [inline pkg/integrations/circleci/on_pipeline_completed.go]: Are we listening to onWorkflowCompleted, or onPipelineCompleted events?

File: pkg/integrations/circleci/trigger_pipeline.go [hunk 4]
```diff
@@ -0,0 +1,541 @@
...
+
+func (t *TriggerPipeline) Configuration() []configuration.Field {
+	return []configuration.Field{
+		{
+			Name:        "projectSlug",
+			Label:       "Project Slug",
+			Type:        configuration.FieldTypeString,
+			Required:    true,
+			Description: "CircleCI project slug (e.g., gh/username/repo)",
+			Placeholder: "gh/username/repo",
```
Comment (shiroyasha) [inline pkg/integrations/circleci/trigger_pipeline.go]: Can we use a resource list for this config field?

File: pkg/integrations/circleci/run_pipeline.go [hunk 5]
```diff
@@ -0,0 +1,541 @@
...
+	})
+
+	if err != nil {
+		return fmt.Errorf("error setting metadata: %v", err)
+	}
+
+	// Request webhook for this project
+	return ctx.Integration.RequestWebhook(WebhookConfiguration{
+		ProjectSlug: config.ProjectSlug,
+	})
```
Comment (shiroyasha) [inline pkg/integrations/circleci/run_pipeline.go:288]: What happens if both the branch and the tag are configured?
Is it a valid configuration, or we need to verify it in the setup.

File: pkg/integrations/circleci/example.go [hunk 6]
```diff
@@ -0,0 +1,44 @@
+package circleci
+
+func (c *TriggerPipeline) ExampleOutput() map[string]any {
```
Comment (lucaspin) [inline pkg/integrations/circleci/example.go]: Let's use json files to define these. Also, event payloads emitted have three field: type, timestamp and data. The payload defined here would be inside the `data` field of the event emitted

File: pkg/integrations/circleci/on_pipeline_completed.go [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+	if err != nil {
+		return fmt.Errorf("failed to decode configuration: %w", err)
+	}
+
+	if config.ProjectSlug == "" {
+		return fmt.Errorf("projectSlug is required")
+	}
+
+	// If this is the same project, nothing to do
+	if metadata.Project != nil && config.ProjectSlug == metadata.Project.Slug {
```
Comment (lucaspin) [inline pkg/integrations/circleci/on_pipeline_completed.go]: We should verify if the project exists here

File: pkg/integrations/circleci/run_pipeline.go [hunk 8]
```diff
@@ -0,0 +1,541 @@
...
+			return http.StatusOK, nil
+		}
+	}
+
+	// All workflows done, emit result
+	if anyFailed {
+		err = executionCtx.ExecutionState.Emit(FailedOutputChannel, PayloadType, []any{payload})
+	} else {
+		err = executionCtx.ExecutionState.Emit(SuccessOutputChannel, PayloadType, []any{payload})
+	}
```
[the remark this answers is not in the document]
Comment (shiroyasha) [reply pkg/integrations/circleci/run_pipeline.go]: @vikramships this seems like a real concern if true. We should return the same payload in both cases.

File: pkg/integrations/circleci/trigger_pipeline.go [hunk 9]
```diff
@@ -0,0 +1,541 @@
...
+		},
+		{
+			Name:        "branch",
```
Comment (lucaspin) [inline pkg/integrations/circleci/trigger_pipeline.go]: We should use the FieldTypeGitRef for the ref, as we do in the Semaphore one, for consistency

File: pkg/integrations/circleci/trigger_pipeline.go [hunk 10]
```diff
@@ -0,0 +1,541 @@
...
+	if err != nil {
+		return fmt.Errorf("failed to decode metadata: %w", err)
+	}
+
+	// If this is the same project, nothing to do
+	if metadata.ProjectSlug == config.ProjectSlug {
+		return nil
+	}
+
+	err = ctx.Metadata.Set(TriggerPipelineNodeMetadata{
```
Comment (lucaspin) [inline pkg/integrations/circleci/trigger_pipeline.go]: We should verify if the project exists and store more information about it in metadata

File: pkg/integrations/circleci/run_pipeline.go [hunk 11]
```diff
@@ -0,0 +1,541 @@
...
+	anyFailed = false
+
+	for _, w := range workflows {
+		if w.Status == "running" || w.Status == "on_hold" || w.Status == "not_run" {
+			allDone = false
+		}
+		if w.Status == WorkflowStatusFailed || w.Status == WorkflowStatusCanceled {
+			anyFailed = true
+		}
+	}
```
[the remark this answers is not in the document]
Comment (shiroyasha) [reply pkg/integrations/circleci/run_pipeline.go]: @vikramships please check if this is a valid concern.

File: pkg/integrations/circleci/circleci.go [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+		return nil, fmt.Errorf("error getting webhook secret: %v", err)
+	}
+
+	// Create CircleCI webhook
+	webhook, err := client.CreateWebhook(
+		name,
+		ctx.Webhook.GetURL(),
+		string(webhookSecret),
+		configuration.ProjectSlug,
+		[]string{"workflow-completed"},
```
Comment (lucaspin) [inline pkg/integrations/circleci/circleci.go]: Shouldn't we also include this in the webhook configuration? What other types of events do we have available?

File: pkg/integrations/circleci/trigger_pipeline.go [hunk 13]
```diff
@@ -0,0 +1,541 @@
...
+	"github.com/superplanehq/superplane/pkg/crypto"
+)
+
+const PayloadType = "circleci.workflow.completed"
+const SuccessOutputChannel = "success"
+const FailedOutputChannel = "failed"
+const WorkflowStatusSuccess = "success"
+const WorkflowStatusFailed = "failed"
+const WorkflowStatusCanceled = "canceled"
+const PollInterval = 10 * time.Second
```
Comment (lucaspin) [inline pkg/integrations/circleci/trigger_pipeline.go]: Polling interval here is too small. Polling should be just a fallback mechanism here in case the webhook fails, so we should use a higher polling interval like 5min

File: pkg/integrations/circleci/run_pipeline.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (lucaspin) [reply pkg/integrations/circleci/run_pipeline.go]: This is a good point. The output should be consistent

File: docs/components/CircleCI.mdx [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+
+<CardGrid>
+  <LinkCard title="Trigger Pipeline" href="#trigger-pipeline" description="Trigger a CircleCI pipeline and wait for completion" />
+</CardGrid>
+
+<a id="on-pipeline-completed"></a>
+
+## On Pipeline Completed
+
+The On Pipeline Completed trigger starts a workflow execution when a CircleCI workflow completes.
```
Comment (shiroyasha) [inline docs/components/CircleCI.mdx]: @vikramships please check if this sentence is correct.

File: pkg/integrations/circleci/trigger_pipeline.go (hunk 13 above)
Comment (shiroyasha) [inline pkg/integrations/circleci/trigger_pipeline.go]: The polling interval is far more aggressive than other integrations (GitHub/ Semaphore are 5 minutes). Is the higher frequency necessary, or should it be relaxed to reduce API pressure and rate limits?

File: docs/components/CircleCI.mdx [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+sidebar:
+  order: 15
+---
+
+Trigger and monitor CircleCI pipelines
+
+## Triggers
+
+<CardGrid>
+  <LinkCard title="On Pipeline Completed" href="#on-pipeline-completed" description="Listen to CircleCI workflow completion events" />
```
Comment (shiroyasha) [inline docs/components/CircleCI.mdx]: The description here is a bit weird/inconsistent:

```
On Pipeline Completed - Listen to CircleCI workflow completion events
```

File: pkg/integrations/circleci/run_pipeline_test.go [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+		for _, param := range params {
+			result[param.Name] = param.Value
+		}
+		result["SUPERPLANE_EXECUTION_ID"] = mockCtx.ID
+		result["SUPERPLANE_CANVAS_ID"] = mockCtx.WorkflowID
+
+		assert.Equal(t, "production", result["env"])
+		assert.Equal(t, "1.0.0", result["version"])
+		assert.Equal(t, "exec-123", result["SUPERPLANE_EXECUTION_ID"])
+		assert.Equal(t, "canvas-456", result["SUPERPLANE_CANVAS_ID"])
```
Comment (shiroyasha) [inline pkg/integrations/circleci/run_pipeline_test.go]: This test on line 106 and 107 are a bit suspicious.

**Comment by shiroyasha** [comment]
@vikramships great start! 🎉 

I left a few comments that need to be addressed before we can merge.

**Comment by forestileao** [comment]
@vikramships thank you for the PR.
Just a few requirements:
- To make it mergable we need to follow some desings on the frontend side. As you can see the component UI is plain with default states and no metadata info. So I suggest checking docs/contributing/component-design.md
- it is possible to use AI help to generate the metadata
- everything that would need to be implemented in the frontend is under web_src/src/worflowv2/mappers
- we have some pre-implemented examples to follow

**Comment by afoofaa** [comment]
> @vikramships thank you for the PR. Just a few requirements:
> 
> * To make it mergable we need to follow some desings on the frontend side. As you can see the component UI is plain with default states and no metadata info. So I suggest checking docs/contributing/component-design.md
> * it is possible to use AI help to generate the metadata
> * everything that would need to be implemented in the frontend is under web_src/src/worflowv2/mappers
> * we have some pre-implemented examples to follow

Hi @forestileao,

I'm working on these improvements based on the examples. Thanks.

**Comment by afoofaa** [comment]
@forestileao @shiroyasha @lucaspin - All feedback addressed. Frontend mappers implemented. Ready for re-review when you get a chance. Thanks!

**Comment by shiroyasha** [comment]
@vikramships it seems that you have conflicts with the main branch, can you please address them 🙏

**Comment by afoofaa** [comment]
> @vikramships it seems that you have conflicts with the main branch, can you please address them 🙏
  
  Thanks for mentioning this @shiroyasha, I pushed new changes.

**Comment by forestileao** [comment]
@vikramships thanks for adding the mapper in the frontend. But please address the comments from @lucaspin and @shiroyasha and resolve them once done.

**Comment by afoofaa** [comment]
Fixed the remaining issues in commit 0131310. Now using JSON files for examples and changed project selection to use dropdown instead of text input. All tests passing. Thanks!

File: pkg/integrations/circleci/list_resources.go (hunk 1 above)
Comment (afoofaa) [reply pkg/integrations/circleci/list_resources.go]: Added ListResources() implementation. Fixed in commit 0131310.

File: pkg/integrations/circleci/on_pipeline_completed.go (hunk 2 above)
Comment (afoofaa) [reply pkg/integrations/circleci/on_pipeline_completed.go]: Changed to FieldTypeIntegrationResource with dropdown. Fixed in commit 0131310.

File: pkg/integrations/circleci/trigger_pipeline.go (hunk 4 above)
Comment (afoofaa) [reply pkg/integrations/circleci/trigger_pipeline.go]: Changed to FieldTypeIntegrationResource. Fixed in commit 0131310.

File: pkg/integrations/circleci/run_pipeline.go (hunk 5 above)
Comment (afoofaa) [reply pkg/integrations/circleci/run_pipeline.go:288]: Added validation that returns error if both are specified. Fixed in earlier commit.

File: pkg/integrations/circleci/example.go (hunk 6 above)
Comment (afoofaa) [reply pkg/integrations/circleci/example.go]: Created JSON files and updated example.go to use embed pattern. Fixed in commit 0131310.

File: pkg/integrations/circleci/on_pipeline_completed.go (hunk 7 above)
Comment (afoofaa) [reply pkg/integrations/circleci/on_pipeline_completed.go]: Added project verification in Setup(). Fixed in earlier commit.

File: pkg/integrations/circleci/trigger_pipeline.go (hunk 9 above)
Comment (afoofaa) [reply pkg/integrations/circleci/trigger_pipeline.go]: Changed to FieldTypeGitRef for both branch and tag. Fixed in earlier commit.

File: pkg/integrations/circleci/trigger_pipeline.go (hunk 10 above)
Comment (afoofaa) [reply pkg/integrations/circleci/trigger_pipeline.go]: Added project verification and store ID, Slug, Name. Fixed in earlier commit.

File: pkg/integrations/circleci/circleci.go (hunk 12 above)
Comment (afoofaa) [reply pkg/integrations/circleci/circleci.go]: Currently scoped to workflow-completed events. Job-completed would need separate trigger. Can expand if needed.

File: pkg/integrations/circleci/trigger_pipeline.go (hunk 13 above)
Comment (afoofaa) [reply pkg/integrations/circleci/trigger_pipeline.go]: Changed to 5 minutes. Fixed in earlier commit.

Comment (afoofaa) [reply pkg/integrations/circleci/trigger_pipeline.go]: Changed to 5 minutes. Fixed in earlier commit.

File: pkg/integrations/circleci/on_workflow_completed_test.go [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+		assert.ErrorContains(t, err, "invalid signature")
+	})
+
+	t.Run("valid signature with workflow-completed event -> event is emitted", func(t *testing.T) {
+		body := []byte(`{"type":"workflow-completed","workflow":{"id":"wf-123","status":"success"},"pipeline":{"id":"pipe-123"}}`)
+
+		secret := "test-secret"
+		h := hmac.New(sha256.New, []byte(secret))
+		h.Write(body)
+		signature := fmt.Sprintf("%x", h.Sum(nil))
```
[the remark this answers is not in the document]
Comment (shiroyasha) [reply pkg/integrations/circleci/on_workflow_completed_test.go:51]: @vikramships can you check this one. Sounds dangerous.

File: pkg/integrations/circleci/circleci.go (hunk 12 above)
Comment (shiroyasha) [reply pkg/integrations/circleci/circleci.go]: @lucaspin this is hard for me to judge. Can you give me an example how this should look like?

**Comment by shiroyasha** [comment]
@vikramships awsm! I see that two things are still failing on CI:

CleanShot 2026-02-07 at 19 14 34@2x

File: pkg/integrations/circleci/circleci.go (hunk 12 above)
Comment (lucaspin) [reply pkg/integrations/circleci/circleci.go]: > Currently scoped to workflow-completed events. Job-completed would need separate trigger. Can expand if needed.

Yes, but in order to extend it without requiring updating the existing database webhook records created for `onPipelineCompleted` triggers or having fallback logic, you need to move this configuration for event types to the webhook configuration. Example, [WebhookConfiguration](https://github.com/superplanehq/superplane/blob/main/pkg/integrations/pagerduty/pagerduty.go#L344) for PagerDuty has an `events` field on it, which allows the [trigger to select to which event types it needs](https://github.com/superplanehq/superplane/blob/main/pkg/integrations/pagerduty/on_incident.go#L167), allowing the [`SetupWebhook()` call](https://github.com/superplanehq/superplane/blob/main/pkg/integrations/pagerduty/pagerduty.go#L421) to not have hardcoded event types on it

Comment (lucaspin) [reply pkg/integrations/circleci/circleci.go]: Also, we use that field in CompareWebhookConfig() to decide if we need to reuse a webhook or not: https://github.com/superplanehq/superplane/blob/main/pkg/integrations/pagerduty/pagerduty.go#L372

**Comment by afoofaa** [comment]
Fixed both. Component docs regenerated and frontend formatted. Tests passing now.

**Comment by afoofaa** [comment]
Fixed the event configuration stuff in commit 7c40589. 

Changes:
• Events field in WebhookConfiguration (not hardcoded anymore)
• Normalizes empty/nil to default ["workflow-completed"] for backward compat  
• Validates against allowed types (workflow-completed, job-completed)
• Dedupes if there are duplicates
• CompareWebhookConfig checks events now (superset matching for webhook reuse)
• TriggerPipeline webhook handler filters by event type before processing
• Added test coverage for all the edge cases

All tests passing, docs regenerated.

File: pkg/integrations/circleci/circleci.go (hunk 12 above)
Comment (afoofaa) [reply pkg/integrations/circleci/circleci.go]: Done. Added Events field and moved event types to WebhookConfiguration. Both triggers now pass events when calling RequestWebhook.

**Comment by afoofaa** [comment]
Fixed the webhook signature verification issue. Now verifying signature before checking event type, same as other integrations. Updated in commit 1036666.

**Comment by afoofaa** [comment]
Rebased onto latest main. Branch is now up to date.

**Comment by shiroyasha** [comment]
@vikramships looks superb. 

Feel free to claim your bounty. @AleksandarCole will help you in the process.

I'll take over this PR for a few commits and touchups, and merge it.

Thank you for contributing to SuperPlane. <3

**Comment by AleksandarCole** [comment]
@vikramships Bounty approved and paid. Thanks for contributing!

**Comment by afoofaa** [comment]
Thank you @shiroyasha and @AleksandarCole for the opportunity and for the smooth review process! 

It was a great learning experience working on this integration. Happy to contribute more to SuperPlane in the future. 🚀

**Comment by shiroyasha** [comment]
@lucaspin @forestileao @AleksandarCole 

I did a couple of changes.

**1/ I reverted back to asking for Project Slug directly instead of listing it in the API.** 

In CircleCI API v2, they don't have a clear/stable way to list projects and the behavior is different depending on how the project is set up (gh oauth, gh app, gitlab, etc...). The solution that I offered up in the comments (using the /me endpoint) is not good enough.

**2/ Fixed icons in the sidebar**

**3/ Pipeline Definition ID**

CircleCI is behaving differently depending on the type of the project. The newer ones are requiring entering a so called "pipeline definition ID" that is displayed in the project settings. Added it to the triggerPipeline.

---

Leftover things:

- [ ] There is a so-called "Pipeline Trigger" API that we could use. It creates an explicit trigger on Circle. Verify this.
- [ ] Verify if we can list the pipeline definition IDs with a dropdown.
- [ ] The name `triggerPipeline` is weird. On most other places we use `runPipeline`. Trigger implies only starting things.

File: pkg/core/component.go [hunk 18]
```diff
@@ -166,6 +168,32 @@ type HTTPContext interface {
...
+	defer res.Body.Close()
+	responseBody, err := io.ReadAll(res.Body)
+	if err != nil {
+		return nil, fmt.Errorf("reading body: %w", err)
+	}
+	if res.StatusCode < 200 || res.StatusCode >= 300 {
+		return nil, fmt.Errorf("request got %d: %s", res.StatusCode, string(responseBody))
+	}
+	return responseBody, nil
+}
```
Comment (shiroyasha) [inline pkg/core/component.go]: 🙈

File: pkg/integrations/circleci/run_pipeline.go [hunk 19]
```diff
@@ -0,0 +1,567 @@
...
+			branch = strings.TrimPrefix(spec.Location, "refs/heads/")
+			params.Branch = branch
+		case strings.HasPrefix(spec.Location, "ref/heads/"):
+			branch = strings.TrimPrefix(spec.Location, "ref/heads/")
+			params.Branch = branch
+		default:
+			branch = strings.TrimSpace(spec.Location)
+			params.Branch = branch
+		}
+	}
```
Comment (shiroyasha) [inline pkg/integrations/circleci/run_pipeline.go:309]: Simplify this.

File: pkg/integrations/circleci/run_pipeline.go [hunk 20]
```diff
@@ -0,0 +1,567 @@
...
+			if branch == "" {
+				branch = "main"
+			}
+			runParams.Config = map[string]string{"branch": branch}
+			runParams.Checkout = map[string]string{"branch": branch}
+		}
+		response, err = client.TriggerPipelineRun(spec.ProjectSlug, runParams)
+	} else {
+		response, err = client.TriggerPipeline(spec.ProjectSlug, params)
+	}
```
Comment (shiroyasha) [inline pkg/integrations/circleci/run_pipeline.go:316]: Too complex, simplify it.

File: pkg/public/server.go [hunk 21]
```diff
@@ -835,7 +835,14 @@ func (s *Server) executeComponentNode(ctx context.Context, body []byte, headers
...
 				Notifications:  contexts.NewNotificationContext(tx, uuid.Nil, execution.WorkflowID),
-			}, nil
+			}
+			if node.AppInstallationID != nil {
+				instance, err := models.FindUnscopedIntegrationInTransaction(tx, *node.AppInstallationID)
+				if err == nil {
+					execCtx.Integration = contexts.NewIntegrationContext(tx, &node, instance, s.encryptor, s.registry)
+				}
+			}
+			return execCtx, nil
```
Comment (shiroyasha) [inline pkg/public/server.go]: Remove this.

**Comment by shiroyasha** [comment]
Continuing tomorrow...

**Comment by omarsoufiane** [comment]
Hello @vikramships,
can you go to bountyhub's dashboard, then payouts, then click on connect stripe to onboard? It's not a vendor stripe account like you'll use on a website, it's a hosted (connected) stripe account on bountyhub and it supports much more countries than the regular stripe. Paypal is available only if your country is not supported by stripe connect, and transaction fees on paypal are higher.

**Comment by afoofaa** [comment]
Hi @omarsoufiane, thanks for the heads up\! I've connected my Stripe account through the BountyHub dashboard. All set now. 🙏

**Comment by omarsoufiane** [comment]
@vikramships, transaction initiated, it should arrive at your bank within 2 to 5 days
